elytron/WFCORE-4750-regex-role-mapper.adoc

```diff
@@ -0,0 +1,109 @@
+
+There should be option to replace all or replace only the first occurrence of the pattern in role. With this parameter we can replace all substrings with another substring. Default is to replace only the first occurrence.
+
+=== Nice-to-Have Requirements
+
+== Implementation Plan
+
+The task will be achieved by implementing new `regex-role-mapper` component in the subsystem (there are already other provided role mappers in the Elytron subsystem, eg. add-prefix-role-mapper, logical-role-mapper etc.). Below is the proposed format of the mapper with examples of use.
+
+* In the example below we will translate all `app-admin`, `batch-admin`, `application-admin` roles to simpler role `admin` and all roles `app-operator`, `batch-operator`, `application-operator` to role `operator`. Roles that do not match the regex will not be kept.
+
+[source,xml]
+----
+<mappers>
+...
+	<regex-role-mapper name="my-regex-role-mapper" pattern=".*-([a-z]*)$" replacement="$1" keep-non-mapped="false"/>
+...
+</mappers>
+
+----
+
+* Customer has roles loaded from LDAP in the following format: `APP-123_XY_ZX_ABCD-Batch_Admin` , `APP-ABC_EF_GH_IJKL-Batch_Operator`. The following mapper can be used to translate these roles to `Admin`, `Operator`:
+
+[source,xml]
+----
+<regex-role-mapper name="my-rrm" regex=".*_([a-zA-Z]*)$" replacement="$1" />
+----
+
+Note: To translate roles `Admin`, `Operator` to `admin`, `operator` with small letter at the beginning, Elytron's `mapping-role-mapper` can be used and chained on top of this one.
+
+* In the example below we will extract roles from the beginning of string before first '_' and transform them to new format APP-{extracted_role} (from input eg. `USER_ABC_DEF`,`ADMIN_123` we will get roles `APP-USER` and `APP-ADMIN`).
+
+[source,xml]
+----
+<regex-role-mapper name="my-rrm" regex="^([a-z]*)_.*" replacement="APP-$1" />
+----
+
+ * In the example below we will extract the domain from roles that are in form of an email and translate them to `{extracted_domain}-role` (eg. from user@gmail.com and user@customeApp.com we will get `gmail-role` and `customerApp-role`). We want to keep the roles that did not match the regex (these were not in form of an email).
+
+[source,xml]
+----
+<regex-role-mapper name="my-rrm" pattern=".*@([a-z]*)\..*" replacement="$1-role" keep-non-mapped="true"/>
+----
+
+ * Role mapper below can be used to replae all instances of `guest` substrings with `user` substring. Eg. role `abc-guest-abc-guest` will be mapped to `abc-user-abc-user`.
+
+[source,xml]
+----
+<regex-role-mapper name="my-rrm" pattern="guest" replacement="user" replace-all="true"/>
+----
+
+Both the pattern and replacement will be in the form of STRING in the model. Validation can take place to notify user if input is not valid regex.
+
+
+== Test Plan
+
+* Functionality tests will be added to Elytron and wildfly-core.
```

Comment:
I see that you didn't include both [1] and [2] sections from the design template. Is there any reason for this? 
I think that - as you provided a tracker for Wildfly documentation, then this should be mentioned in [1] and also that this nice new feature could be worth at least a sentence in the `Release Notes`. At least I - as a developer  - would find it really helpful if it was there :) 
WDYT?

[1]
https://github.com/wildfly/wildfly-proposals/blob/master/design-doc-template.adoc#community-documentation

[2]
https://github.com/wildfly/wildfly-proposals/blob/master/design-doc-template.adoc#release-note-content